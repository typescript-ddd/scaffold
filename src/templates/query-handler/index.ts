export * from "./query-handler.generator";
export * from "./query-handler.template";
export * from "./query-handler.types";
