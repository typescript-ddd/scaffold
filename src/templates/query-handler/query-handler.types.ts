import {
  TemplateBooleanOption,
  TemplatePropertiesOption,
  TemplatePropertyValue,
  TemplateStringOption,
} from "../shared/template";

export type QueryHandlerTemplateValues = {
  entityName: string;
  actionName?: string;
  actor: string;
  returnType: string;
  queryProperties: TemplatePropertyValue[];
  returnsView: boolean;
};

export type QueryHandlerTemplateOptions = {
  entityName: TemplateStringOption;
  actionName?: TemplateStringOption;
  actor: TemplateStringOption;
  returnType: TemplateStringOption;
  queryProperties: TemplatePropertiesOption;
  returnsView: TemplateBooleanOption;
};
