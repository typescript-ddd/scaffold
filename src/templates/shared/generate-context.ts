import { SourceFile } from "ts-morph";

export type ContextImportDeclaration = { moduleSpecifier: string; namedImports: string[] };

export interface GenerateContext {
    resolveDir(type: "domain" | "infrastructure" | "application" | "presentation", subPath?: string): string;
    fileName: string;
    addImportDeclaration(declaration: ContextImportDeclaration): void;
    getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[];
    getModuleSpecifiers(): string[];
    getNamedImports(moduleSpecifier?: string): string[];    
    writeSourceFileImports: (sourceFile: SourceFile) => void;
    clearImportDeclarations(): void;
    tokenize: (text: string, tokens: { name: string }[]) => string;
}