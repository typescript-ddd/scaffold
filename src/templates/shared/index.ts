export * from "./context-builder";
export * from "./generate-context";