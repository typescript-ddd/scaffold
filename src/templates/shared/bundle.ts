import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
