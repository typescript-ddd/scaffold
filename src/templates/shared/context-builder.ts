import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
