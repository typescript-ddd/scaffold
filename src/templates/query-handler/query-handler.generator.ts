import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { QueryHandlerTemplateValues } from "./query-handler.types";

const actorMap: Record<string, string> = {
  find: "finder",
};

export const generateQueryHandler = (
  values: QueryHandlerTemplateValues,
  context: GenerateContext
) => {
  const {
    entityName,
    actionName = "find",
    actor,
    queryProperties,
    returnType,
    returnsView = false,
  } = values;
  const entityType = strings.capitalize(entityName);
  const actionType = strings.capitalize(actionName);
  const actorType = strings.capitalize(actor);
  const queryType = `${actionType}${entityType}Query`;
  const actorName = strings.lower(actorMap[actionName] ?? "finder");

  const className = `${actionType}${entityType}Handler`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} implements QueryHandler<${queryType}> {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["QueryHandler"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("application"),
    namedImports: [queryType, actorType],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(className)}.`,
  });

  classDeclaration
    .addConstructor({
      parameters: [
        {
          name: actorName,
          type: actorType,
          isReadonly: true,
          scope: Scope.Private,
        },
      ],
    })
    .addJsDoc({
      description: `Initializes a new instance of ${strings.describe(
        className
      )}.`,
      tags: [
        {
          tagName: "param",
          text: `{${actorType}} ${actorName} - The ${actorName} to handle the query.`,
        },
      ],
    });

  classDeclaration
    .addMethod({
      name: "handle",
      isAsync: true,
      returnType: `Promise<${returnType}>`,
      parameters: [{ name: "query", type: queryType }],
    })
    .setBodyText((writer) => {
      const isArray = returnType.endsWith("[]");
      const returnName = isArray ? strings.plural(strings.lower(entityName)) : strings.lower(entityName);

      writer.writeLine(
        `const ${returnName} = await this.${actorName}.${strings.camel(actionName)}(`
      );
      queryProperties.forEach((property, index) => {
        writer.writeLine(
          `  query.${property.name}${
            index === queryProperties.length - 1 ? "" : ","
          }`
        );
      });
      writer.writeLine(`);`);
      if (!returnsView) {
        writer.writeLine(`return ${returnName};`);
      } else {
        const viewCreator = returnType.endsWith("[]") ? returnType.slice(0, -2) : returnType;
        writer.writeLine(`return ${viewCreator}.${isArray ? "createMany" : "create"}(${strings.lower(entityName)});`);
      }
    })
    .addJsDoc({
      description: `Handles ${strings.describe(queryType)}.`,
      tags: [
        {
          tagName: "param",
          text: `{${queryType}} query - The query to handle.`,
        },
        {
          tagName: "returns",
          text: `{Promise<${returnType}>} - ${strings.capitalize(
            strings.describe(returnType)
          )}.`,
        },
      ],
    });

  let output = sourceFile.getFullText();

  return output;
};
