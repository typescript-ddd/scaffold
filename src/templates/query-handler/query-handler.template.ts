import { Chunk, GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateQueryHandler } from "./query-handler.generator";
import {
  QueryHandlerTemplateOptions,
  QueryHandlerTemplateValues,
} from "./query-handler.types";

export class QueryHandlerTemplate
  implements Template<QueryHandlerTemplateValues, QueryHandlerTemplateOptions>
{
  readonly name = "Query Handler";
  readonly description = "Generates a query handler";
  readonly options: QueryHandlerTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
    actionName: {
      name: "action",
      label: "Action",
      description: "The action of the query handler",
      optionType: "string",
    },
    actor: {
      name: "actor",
      label: "Actor",
      description: "The actor of the query handler",
      optionType: "string",
    },
    queryProperties: {
      name: "queryProperties",
      label: "Query Properties",
      description: "The query properties of the query handler",
      optionType: "properties",
    },
    returnType: {
      name: "returnType",
      label: "Return Type",
      description: "The return type of the query handler",
      optionType: "string",
    },
    returnsView: {
      name: "returnsView",
      label: "Returns View",
      description: "Whether the query handler returns a view",
      optionType: "boolean",
    },
  };
  readonly defaultValues = {
    entityName: "User",
    actionName: "findAll",
    actor: "UserFinder",
    returnType: "UserView[]",
    returnsView: true,
    queryProperties: [{ name: "name", valueType: "string" }],
  };
  generate(values: QueryHandlerTemplateValues, context: GenerateContext): Chunk {
    return generateQueryHandler(values, context);
  }
}
