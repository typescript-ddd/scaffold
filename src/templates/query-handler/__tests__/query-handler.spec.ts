import { ContextBuilder, GenerateContext } from "../../shared";
import { generateQueryHandler } from "../query-handler";

describe("CommandHandler", () => {
    let context: GenerateContext;

    beforeEach(() => {
        context = ContextBuilder.build("@src");
    })

    it("Should render the command handler template", () => {
        
        context.addImportDeclaration({
            moduleSpecifier: context.resolveDir("domain"),
            namedImports: ["UserView"],
        });

        const output = generateQueryHandler(
            {
                entityName: "User",
                actor: "UserFinder",
                returnType: "UserView[]",
                actionName: "findAll",
                returnsView: true,
                queryProperties: [
                    { name: "name", type: "string" },
                ]
            },
            context
        );

        expect(output).toMatchSnapshot();
    });
})