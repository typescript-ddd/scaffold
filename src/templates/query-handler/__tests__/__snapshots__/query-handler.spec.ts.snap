// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandHandler Should render the command handler template 1`] = `
"import { QueryHandler } from "@typescript-ddd/core";
import { UserView } from "@src/domain";
import { FindUserQuery, UserFinder } from "@src/application";

/** Represents a find user handler. */
export class FindUserHandler implements QueryHandler<FindUserQuery> {
    /**
     * Initializes a new instance of a find user handler.
     * @param {UserFinder} finder - The finder to handle the query.
     */
    constructor(private readonly finder: UserFinder) {
    }

    /**
     * Handles a find user query.
     * @param {FindUserQuery} query - The query to handle.
     * @returns {Promise<UserView>} - A user view.
     */
    async handle(query: FindUserQuery): Promise<UserView> {
        const user = await this.finder.find(
          query.name
        );
        return UserView.find(user);
    }
}    
"
`;
